src/main/java/carpet/mixins/PistonBlock_movableTEMixin.java

```diff
@@ -0,0 +1,123 @@
+        {
+            return !true;
+        }
+        else
+        {
+            return !(CarpetSettings.b_movableTileEntities && isPushableTileEntityBlock(block));
+        }
+    }
+    @Inject(method = "move", at = @At(value = "INVOKE", shift = At.Shift.BEFORE,
+            target = "Ljava/util/List;size()I", ordinal = 4),locals = LocalCapture.CAPTURE_FAILHARD)
+    private void onMove(World world_1, BlockPos blockPos_1, Direction direction_1, boolean boolean_1,
+            CallbackInfoReturnable<Boolean> cir, BlockPos blockPos_2, PistonHandler pistonHandler_1,
+            List<BlockPos> list_1, List<BlockState> list_2, List list_3, int int_2, BlockState blockStates_1[],
+            Direction direction_2, Set set_1)
+    {
+        //Get the blockEntities and remove them from the world before any magic starts to happen
+        if (CarpetSettings.b_movableTileEntities)
+        {
+            list1_BlockEntities.set(Lists.<BlockEntity>newArrayList());
+            for (int i = 0; i < list_1.size(); ++i)
+            {
+                BlockPos blockpos = list_1.get(i);
+                BlockEntity blockEntity = (list_2.get(i).getBlock().hasBlockEntity()) ? world_1.getBlockEntity(blockpos) : null;
+                list1_BlockEntities.get().add(blockEntity);
+                if (blockEntity != null)
+                {
+                    //hopefully this call won't have any side effects in the future, such as dropping all the BlockEntity's items
+                    //we want to place this same(!) BlockEntity object into the world later when the movement stops again
+                    world_1.removeBlockEntity(blockpos);
+                    blockEntity.markDirty();
+                }
+            }
+        }
+    }
+    
+    @Inject(method = "move", at = @At(value = "INVOKE", shift = At.Shift.BEFORE,
+            target = "Lnet/minecraft/world/World;setBlockEntity(Lnet/minecraft/util/math/BlockPos;" +
+                             "Lnet/minecraft/block/entity/BlockEntity;)V", ordinal = 0),
+            locals = LocalCapture.CAPTURE_FAILHARD)
+    private void setBlockEntityWithCarried(World world_1, BlockPos blockPos_1, Direction direction_1, boolean boolean_1,
+            CallbackInfoReturnable<Boolean> cir, BlockPos blockPos_2, PistonHandler pistonHandler_1, List list_1,
+            List list_2, List list_3, int int_2, BlockState blockStates_1[], Direction direction_2, Set set_1,
+            int int_3, BlockPos blockPos_4, BlockState blockState_1)
+    {
+        BlockEntity blockEntityPiston = PistonExtensionBlock.createBlockEntityPiston((BlockState) list_2.get(int_3),
+                direction_1, boolean_1, false);
+        if (CarpetSettings.b_movableTileEntities)
+            ((PistonBlockEntityInterface) blockEntityPiston).setCarriedBlockEntity(list1_BlockEntities.get().get(int_3));
+        world_1.setBlockEntity(blockPos_4, blockEntityPiston);
+    }
+    
+    @Redirect(method = "move", at = @At(value = "INVOKE",
```

Comment:
OK, so you cancel vanilla call here. I assume that only the trick to being able to capture locals while doing a redirect, correct?

Reply:
Yes, unfortunately